File: services/api/src/main/java/com/launchly/testcase/services/TestService.java
package com.launchly.testcase.services;

import com.launchly.testcase.dto.*;
import com.launchly.testcase.entities.TestCase;
import com.launchly.testcase.entities.TestRun;
import com.launchly.testcase.entities.TestRunCase;
import com.launchly.testcase.enums.Priority;
import com.launchly.testcase.enums.TestCaseStatus;
import com.launchly.testcase.enums.TestResult;
import com.launchly.testcase.enums.TestRunStatus;
import com.launchly.testcase.repositories.TestCaseRepository;
import com.launchly.testcase.repositories.TestRunCaseRepository;
import com.launchly.testcase.repositories.TestRunRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TestService {
    private final TestCaseRepository testCaseRepository;
    private final TestRunRepository testRunRepository;
    private final TestRunCaseRepository testRunCaseRepository;

    public TestService(TestCaseRepository testCaseRepository,
                       TestRunRepository testRunRepository,
                       TestRunCaseRepository testRunCaseRepository) {
        this.testCaseRepository = testCaseRepository;
        this.testRunRepository = testRunRepository;
        this.testRunCaseRepository = testRunCaseRepository;
    }

    // --- TestCase ---
    public TestCaseResponse createTestCase(String projectId, TestCaseRequest request) {
        TestCase tc = new TestCase();
        tc.setProjectId(projectId);
        tc.setTitle(request.title());
        tc.setModule(request.module());
        tc.setSteps(request.steps());
        tc.setExpectedResult(request.expectedResult());
        tc.setPriority(request.priority() != null ? Priority.valueOf(request.priority()) : Priority.P2);
        tc.setTags(request.tags());
        tc.setOwnerId(request.ownerId());
        return TestCaseResponse.from(testCaseRepository.save(tc));
    }

    public List<TestCaseResponse> listTestCases(String projectId) {
        return testCaseRepository.findByProjectIdOrderByCreatedAtDesc(projectId)
                .stream().map(TestCaseResponse::from).collect(Collectors.toList());
    }

    public TestCaseResponse getTestCase(String id) {
        return TestCaseResponse.from(testCaseRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("TestCase not found: " + id)));
    }

    public TestCaseResponse updateTestCase(String id, TestCaseRequest request) {
        TestCase tc = testCaseRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("TestCase not found: " + id));
        tc.setTitle(request.title());
        tc.setModule(request.module());
        tc.setSteps(request.steps());
        tc.setExpectedResult(request.expectedResult());
        if (request.priority() != null) tc.setPriority(Priority.valueOf(request.priority()));
        tc.setTags(request.tags());
        tc.setOwnerId(request.ownerId());
        return TestCaseResponse.from(testCaseRepository.save(tc));
    }

    public void deleteTestCase(String id) {
        testCaseRepository.deleteById(id);
    }

    // --- TestRun ---
    @Transactional
    public TestRunResponse createTestRun(String deploymentId, String projectId, String environmentId, String userId) {
        TestRun tr = new TestRun();
        tr.setDeploymentId(deploymentId);
        tr.setProjectId(projectId);
        tr.setEnvironmentId(environmentId);
        tr.setCreatedBy(userId);
        tr.setStatus(TestRunStatus.PENDING);
        tr = testRunRepository.save(tr);

        // Auto-create TestRunCase for all ACTIVE test cases
        List<TestCase> activeCases = testCaseRepository.findByProjectIdAndStatus(
                projectId, TestCaseStatus.ACTIVE);
        for (TestCase tc : activeCases) {
            TestRunCase trc = new TestRunCase();
            trc.setTestRunId(tr.getId());
            trc.setTestCaseId(tc.getId());
            trc.setResult(TestResult.SKIPPED);
            testRunCaseRepository.save(trc);
        }

        return TestRunResponse.from(tr);
    }

    public TestRunResponse getTestRun(String id) {
        return TestRunResponse.from(testRunRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("TestRun not found: " + id)));
    }

    public List<TestRunResponse> listTestRuns(String projectId) {
        return testRunRepository.findByProjectIdOrderByCreatedAtDesc(projectId)
                .stream().map(TestRunResponse::from).collect(Collectors.toList());
    }

    public List<TestRunCaseResponse> getTestRunCases(String testRunId) {
        return testRunCaseRepository.findByTestRunId(testRunId)
                .stream().map(TestRunCaseResponse::from).collect(Collectors.toList());
    }

    public TestRunCaseResponse updateTestRunCase(String testRunId, String caseId,
                                                  UpdateTestRunCaseRequest request) {
        return updateTestRunCase(testRunId, caseId, request, request.executedBy());
    }

    public TestRunCaseResponse updateTestRunCase(String testRunId, String caseId,
                                                  UpdateTestRunCaseRequest request, String userId) {
        TestRunCase trc = testRunCaseRepository.findById(caseId)
                .orElseThrow(() -> new IllegalArgumentException("TestRunCase not found: " + caseId));
        if (!trc.getTestRunId().equals(testRunId)) {
            throw new IllegalArgumentException("TestRunCase does not belong to this TestRun");
        }
        if (request.result() != null) {
            trc.setResult(TestResult.valueOf(request.result()));
        }
        trc.setNotes(request.notes());
        trc.setExecutedBy(userId);
        trc.setExecutedAt(Instant.now());
        trc = testRunCaseRepository.save(trc);

        // Check if all cases are done
        List<TestRunCase> allCases = testRunCaseRepository.findByTestRunId(testRunId);
        boolean allDone = allCases.stream().noneMatch(c -> c.getResult() == TestResult.SKIPPED);
        if (allDone) {
            testRunRepository.findById(testRunId).ifPresent(tr -> {
                tr.setStatus(TestRunStatus.COMPLETED);
                tr.setFinishedAt(Instant.now());
                testRunRepository.save(tr);
            });
        }

        return TestRunCaseResponse.from(trc);
    }
}

File: services/api/src/testcase/test.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../common/prisma/prisma.service';

@Injectable()
export class TestService {
  constructor(private readonly prisma: PrismaService) {}

  async createTestCase(projectId: string, data: any) {
    return this.prisma.testCase.create({
      data: {
        projectId,
        title: data.title,
        description: data.description,
        priority: data.priority || 'P2',
        steps: data.steps,
        expectedResult: data.expectedResult,
      },
    });
  }

  async listTestCases(projectId: string) {
    return this.prisma.testCase.findMany({
      where: { projectId },
      orderBy: { createdAt: 'desc' },
    });
  }

  async getTestCase(id: string) {
    const tc = await this.prisma.testCase.findUnique({ where: { id } });
    if (!tc) throw new NotFoundException('Test case not found');
    return tc;
  }

  async updateTestCase(id: string, data: any) {
    return this.prisma.testCase.update({
      where: { id },
      data: {
        ...(data.title !== undefined && { title: data.title }),
        ...(data.description !== undefined && { description: data.description }),
        ...(data.priority !== undefined && { priority: data.priority }),
        ...(data.steps !== undefined && { steps: data.steps }),
        ...(data.expectedResult !== undefined && { expectedResult: data.expectedResult }),
        ...(data.status !== undefined && { status: data.status }),
      },
    });
  }

  async deleteTestCase(id: string) {
    await this.prisma.testCase.delete({ where: { id } });
  }

  async createTestRun(deploymentId: string, projectId: string, environmentId: string, userId: string) {
    const deployment = await this.prisma.deployment.findUnique({
      where: { id: deploymentId },
      select: { id: true, projectId: true, environmentId: true },
    });
    if (!deployment) {
      throw new NotFoundException('部署不存在');
    }
    if (deployment.projectId !== projectId) {
      throw new BadRequestException('部署不属于当前项目');
    }

    const targetEnvironmentId = environmentId || deployment.environmentId;
    if (!targetEnvironmentId) {
      throw new BadRequestException('环境不能为空');
    }
    if (environmentId && environmentId !== deployment.environmentId) {
      throw new BadRequestException('环境不属于当前部署');
    }

    const testCases = await this.prisma.testCase.findMany({
      where: { projectId, status: 'ACTIVE' },
    });

    return this.prisma.$transaction(async (tx) => {
      const run = await tx.testRun.create({
        data: {
          deploymentId,
          projectId,
          environmentId: targetEnvironmentId,
          totalCases: testCases.length,
          triggeredBy: userId,
        },
      });

      if (testCases.length > 0) {
        await tx.testRunCase.createMany({
          data: testCases.map(tc => ({
            testRunId: run.id,
            testCaseId: tc.id,
          })),
        });
      }

      return run;
    });
  }

  async getTestRun(id: string) {
    const run = await this.prisma.testRun.findUnique({ where: { id } });
    if (!run) throw new NotFoundException('Test run not found');
    return run;
  }

  async listTestRuns(projectId: string) {
    return this.prisma.testRun.findMany({
      where: { projectId },
      orderBy: { createdAt: 'desc' },
    });
  }

  async getTestRunCases(testRunId: string) {
    return this.prisma.testRunCase.findMany({
      where: { testRunId },
      include: { testCase: true },
    });
  }

  async updateTestRunCase(testRunId: string, caseId: string, data: any, userId: string) {
    const runCase = await this.prisma.testRunCase.findFirst({
      where: { id: caseId, testRunId },
    });
    if (!runCase) throw new NotFoundException('Test run case not found');

    const updated = await this.prisma.testRunCase.update({
      where: { id: caseId },
      data: {
        result: data.result,
        notes: data.notes,
        executedBy: userId,
        executedAt: new Date(),
      },
    });

    // Update test run counters
    await this.updateTestRunCounters(testRunId);

    return updated;
  }

  private async updateTestRunCounters(testRunId: string) {
    const cases = await this.prisma.testRunCase.findMany({
      where: { testRunId },
    });

    const passed = cases.filter(c => c.result === 'PASSED').length;
    const failed = cases.filter(c => c.result === 'FAILED').length;
    const skipped = cases.filter(c => c.result === 'SKIPPED').length;
    const allDone = cases.every(c => c.result !== 'PENDING');

    await this.prisma.testRun.update({
      where: { id: testRunId },
      data: {
        passedCases: passed,
        failedCases: failed,
        skippedCases: skipped,
        status: allDone ? 'COMPLETED' : 'RUNNING',
        finishedAt: allDone ? new Date() : null,
      },
    });
  }
}
